diff --git non_module_deps.bzl non_module_deps.bzl
new file mode 100644
index 00000000..4c549536
--- /dev/null
+++ non_module_deps.bzl
@@ -0,0 +1,15 @@
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+def _non_module_deps_impl(ctx):
+    http_archive(
+        name = "rules_rust_tinyjson",
+        sha256 = "1a8304da9f9370f6a6f9020b7903b044aa9ce3470f300a1fba5bc77c78145a16",
+        url = "https://crates.io/api/v1/crates/tinyjson/2.3.0/download",
+        strip_prefix = "tinyjson-2.3.0",
+        type = "tar.gz",
+        build_file = "@rules_rust//util/process_wrapper:BUILD.tinyjson.bazel",
+    )
+
+non_module_deps = module_extension(
+    implementation = _non_module_deps_impl,
+)
