diff --git a/java/repositories.bzl b/java/repositories.bzl
index f60a4d0..8b5babd 100644
--- a/java/repositories.bzl
+++ b/java/repositories.bzl
@@ -49,6 +49,8 @@ JAVA_TOOLS_CONFIG = {
             "mirror_url": "https://mirror.bazel.build/bazel_java_tools/releases/java/v13.6.1/java_tools-v13.6.1.zip",
             "github_url": "https://github.com/bazelbuild/java_tools/releases/download/java_13.6.1/java_tools-v13.6.1.zip",
             "sha": "74c978eab040ad4ec38ce0d0970ac813cc2c6f4f6f4f121c0414719487edc991",
+            "patches": ["@@//patches:remote_java_tools.patch"],
+            "patch_args": ["-d", "java_tools/zlib", "-p1"],
         },
     },
 }
@@ -64,6 +66,8 @@ def java_tools_repos():
                 config["mirror_url"],
                 config["github_url"],
             ],
+            patches = config.get("patches", []),
+            patch_args = config.get("patch_args", []),
         )
 
 def local_jdk_repo():
